# Basic Arithmetic and Vectorized Operations (Doing Math Without Loops)

In regular Python, if you want to add `10` to every number in a list, you have to write a loop. In Python, writing `for`
loops is very slow because Python is an interpreted language (it reads and executes code line by line).

NumPy uses **Vectorization** (performing calculations on entire arrays at once). This runs in highly optimized, pre-compiled
C-code under the hood. It utilizes CPU features like **SIMD** (Single Instruction Multiple Data), which applies a mathematical
instruction to multiple pieces of data at the exact same millisecond. This is why NumPy is up to **100 times faster** than plain Python loops!


## 2. Simple Explanation with an Everyday Analogy
Imagine you are a teacher with 30 students in a classroom. You have graded their test papers, but you want to give everyone a **5-point bonus**:
*   **The Loop Method**: You walk up to Student 1, add 5 points to their paper, write it down. Then you walk to Student 2, add 5 points... this takes a long time (slow).
*   **The Vectorized Method**: You use a magical loudspeaker that says, "Everyone, add 5 points to your score right now!" Every student does it at the exact same second (extremely fast).


#### Scalar Arithmetic
A **scalar** is just a single number (like `5` or `10.5`). When you apply a math operator between a NumPy array and a scalar, NumPy automatically applies that operator to **every element** in the array.


In [1]:
import numpy as np

a = np.array([1, 2, 3, 4])

print("a + 10 =", a + 10)
print("a - 2  =", a - 2)
print("a * 3  =", a * 3)
print("a / 2  =", a / 2)
print("a ** 2 =", a ** 2) # Power of 2 (Squaring)

a + 10 = [11 12 13 14]
a - 2  = [-1  0  1  2]
a * 3  = [ 3  6  9 12]
a / 2  = [0.5 1.  1.5 2. ]
a ** 2 = [ 1  4  9 16]


#### Elementwise Arithmetic between Arrays
If you have two arrays of the **same shape**, you can add, subtract, multiply, or divide them. NumPy does this **element-by-element** (first element with first element, second with second, etc.).


In [2]:
import numpy as np

a1 = np.array([1, 2, 3])
a2 = np.array([10, 20, 30])

print("a1 + a2 =", a1 + a2)
print("a1 * a2 =", a1 * a2) # Element-by-element multiplication!
print("a2 / a1 =", a2 / a1)

a1 + a2 = [11 22 33]
a1 * a2 = [10 40 90]
a2 / a1 = [10. 10. 10.]


#### Comparison Operators (Boolean Arrays)
You can use comparison operators like `>`, `<`, `==`, or `>=`. NumPy will compare every element and return an array of `True` and `False` values (called a **boolean array** or **boolean mask**).


In [3]:
import numpy as np

scores = np.array([91, 55, 100, 73, 82])

passing_scores = scores >= 60
print("Scores >= 60:", passing_scores)

# We can even use conditional assignment!
# If score is below 60, change it to 0
scores[scores < 60] = 0
print("Modified scores:", scores)

Scores >= 60: [ True False  True  True  True]
Modified scores: [ 91   0 100  73  82]



#### Vectorized Math Functions (ufuncs)
NumPy includes highly optimized math functions (Universal Functions or **ufuncs**). When you pass an array to these functions, they calculate the result for every element in parallel.
*   `np.sqrt(a)`: Square root of each element.
*   `np.round(a)`: Round to nearest integer.
*   `np.floor(a)`: Round down.
*   `np.ceil(a)`: Round up.
*   `np.pi`: Constant for $\pi$ (approx. 3.14159).

In [4]:
import numpy as np

radii = np.array([1.0, 2.5, 3.99])

print("Square roots:", np.sqrt(radii))
print("Rounded radii:", np.round(radii))
print("Floor (round down):", np.floor(radii))
print("Ceiling (round up):", np.ceil(radii))
print("Pi constant:", np.pi)

Square roots: [1.         1.58113883 1.99749844]
Rounded radii: [1. 2. 4.]
Floor (round down): [1. 2. 3.]
Ceiling (round up): [1. 3. 4.]
Pi constant: 3.141592653589793


 ### Common Mistakes Beginners Make

1.  **Thinking `+` merges arrays like Python lists**:

In [5]:
    list1 = [1, 2]
    list2 = [3, 4]
    print(list1 + list2) # Output: [1, 2, 3, 4] (Merges/Concatenates them)

    arr1 = np.array([1, 2])
    arr2 = np.array([3, 4])
    print(arr1 + arr2) # Output: [4, 6] (Vector Addition!)

[1, 2, 3, 4]
[4 6]


2.  **Expecting `a * b` to perform matrix multiplication**: In linear algebra, multiplying two matrices is a special process (dot product). In NumPy, writing `a * b` simply multiplies matching elements. If you want true matrix multiplication, you must use `a @ b` or `np.matmul(a, b)`.
3.  **Forgetting that operations return a new array**: Math operations like `a + 10` do not change the original array `a` unless you assign it back: `a = a + 10` or use `a += 10`.



#### Exercise
Create a NumPy array with the numbers `[10, 20, 30]`.
1. Multiply every number by `5`.
2. Subtract `15` from the result.
3. Show the final array.

In [6]:
import numpy as np
a = np.array([10, 20, 30])
result = (a * 5) - 15
print(result)

[ 35  85 135]


*Explanation*: NumPy processes the scalar multiplication first: `[50, 100, 150]`. Then it subtracts `15` elementwise: `[35, 85, 135]`.


#### Exercise
You are given two arrays:

In [ ]:
x = np.array([1.5, 2.7, 3.9])
y = np.array([2.0, 2.0, 3.0])

1. Multiply array `x` and `y` elementwise.
2. Apply `np.sqrt()` to the resulting multiplication array.
3. Round the square root values to the nearest integer using `np.round()`.

In [7]:
import numpy as np
x = np.array([1.5, 2.7, 3.9])
y = np.array([2.0, 2.0, 3.0])

# Multiply
mult = x * y
# Square root
s = np.sqrt(mult)
# Round
rounded = np.round(s)

print("Multiplication:", mult)
print("Square root:", s)
print("Rounded:", rounded)

Multiplication: [ 3.   5.4 11.7]
Square root: [1.73205081 2.32379001 3.42052628]
Rounded: [2. 2. 3.]


*Explanation*:
*   `x * y` multiplies matching elements: `[1.5*2.0, 2.7*2.0, 3.9*3.0]` which equals `[3.0, 5.4, 11.7]`.
*   `np.sqrt()` calculates square roots in parallel: e.g., $\sqrt{3.0} \approx 1.732$.
*   `np.round()` rounds `1.732` to `2.0`, `2.324` to `2.0`, and `3.42` to `3.0`.

#### Exercise
You have an array containing the radiuses of 3 spheres:

In [ ]:
r = np.array([1.0, 2.5, 4.0])

Calculate the volume of each sphere using the mathematical formula:
$$Volume = \frac43 \pi r^3$$
Use `np.pi` for $\pi$ and round your final volume answers to **2 decimal places** using `np.round()`.

In [8]:
import numpy as np
r = np.array([1.0, 2.5, 4.0])

# Calculate sphere volumes
volumes = (4/3) * np.pi * (r ** 3)

# Round to 2 decimal places
rounded_volumes = np.round(volumes, decimals=2)
print("Volumes:", rounded_volumes)

Volumes: [  4.19  65.45 268.08]


*Explanation*:
*   `r ** 3` cubes the radiuses: `[1.0, 15.625, 64.0]`.
*   Then we multiply by $\pi$ and by $4/3$ (scalar multiplication).
*   `np.round(volumes, decimals=2)` rounds the resulting float values to exactly two digits after the decimal point.